### PR TITLE
#15: add Flow/TypeScript support (closes #15)

### LINKED ISSUE #15: add Flow/TypeScript support
## requirements
add Flow/TypeScript support

## specs
{optional: describe technical specs to implement this feature, if not obvious}

## misc
{optional: other useful info}

### PR DESCRIPTION
**Pull request by gcanti**
Proof of concept

This PR adds:
- for consumers
  - Flow and TypeScript typings, shipped with the npm package itself
- for developers
  - source code ready for Flow (typings of dependencies that are needed are included)

### CODE REVIEW

File: flow-typed/buildo.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+type Many<T> = T | T[];
```
Comment (giogonzo) [inline flow-typed/buildo.js]: why is this named `buildo.js`?

File: flow-typed/npm/classnames_v2.x.x.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// flow-typed signature: cf6332fcf9a3398cffb131f7da90662b
+// flow-typed version: dc0ded3d57/classnames_v2.x.x/flow_>=v0.28.x
```
Comment (giogonzo) [inline flow-typed/npm/classnames_v2.x.x.js]: This means this is a publicly available typing? Why are we committing this file?

File: package.json [hunk 3]
```diff
@@ -2,10 +2,15 @@
...
+  "files": [
+    "lib",
+    "react-flexview.d.ts"
+  ],
+  "main": "lib/FlexView.js",
+  "typings": "react-flexview.d.ts",
   "scripts": {
     "test": "mocha",
-    "build": "rm -rf lib && mkdir lib && node-sass src --importer sass-importer.js --include-path node_modules -o lib && babel src -d lib",
+    "build": "rm -rf lib && mkdir lib && node-sass src --importer sass-importer.js --include-path node_modules -o lib && babel src -d lib && flow-copy-source -v src lib",
```
Comment (giogonzo) [inline package.json]: why is this "copy" step needed?

File: flow-typed/buildo.js (hunk 1 above)
Comment (giogonzo) [inline flow-typed/buildo.js]: discussing offline:
- this file is for "missing typings", where missing could be either:
  - not part of the package on npm or in any typing repo (applies both to Flow and TypeScript) (the case of `tcomb-react`)
  - incomplete/loose typing available, or, like it is the case for the `lodash` one here: we'd want to limit the `omit` and `pick` apis to something more strict and less funky JS (e.g. required first Object and/or required at least something to "pick"/"omit")
- it is named "buildo" because in theory, such typings could be shared cross-repo, but this is a whole new field to explore

Summing up, this file should probably be renamed (and maybe split), and we should keep just what we deem to be strictly needed to support development (with Flow)

File: flow-typed/npm/classnames_v2.x.x.js (hunk 2 above)
Comment (giogonzo) [inline flow-typed/npm/classnames_v2.x.x.js]: Offline answer: something that should be discussed as well. here we are kind of "pinning" to a specific version of this typing (shielding from possible typing breakages on flow typed). Do we consider this relevant? Should we? to be discussed

File: package.json (hunk 3 above)
Comment (giogonzo) [inline package.json]: since `flow gen-flow-files` is experimental/broken, the easiest way to ship typings along with the package on npm is to copy the full source, including flow type labels

the order of typing resolution for flow goes like:
- types in the package itself
- typings from flow-typed
- fallback to `any`

**Comment by giogonzo** [comment]
@gcanti I also edited the PR description adding an overview

Summing up (you can see more comments inline), I'm in favor of merging this provided a few changes:
- remove `/lib` from git
- decide what to do with `flow-typed/buildo.js`: in my opinion, for now, just keep what's strictly necessary:
  - nothing for `tcomb-react` (isn't `any` the default fallback anyway?)
  - `lodash/pick/omit`
    - if we really want it, move to a `lodash.js` file?
    - if not, maybe just rely on the publicly available typings (which are correct wrt the real api, but more "funky", as the api is, compared to the ones provided here)

/cc @gabro @FrancescoCioria

File: package.json [hunk 4]
```diff
@@ -2,10 +2,15 @@
   "name": "react-flexview",
   "version": "1.0.1",
   "description": "A react component to abstract over flexbox",
-  "main": "index.js",
+  "files": [
+    "lib",
```
Comment (FrancescoCioria) [inline package.json]: If we decide to replace `.npmignore` with `files` we should export also `src` as it's used in many of our projects.

I would do this in another issue anyway

File: src/index.js [hunk 5]
```diff
@@ -1 +0,0 @@
-export default from './FlexView';
```
Comment (FrancescoCioria) [inline src/index.js]: why did you remove this? this is definitely a big breaking change. is it necessary?

File: package.json [hunk 6]
```diff
@@ -2,10 +2,15 @@
...
+    "react-flexview.d.ts"
+  ],
+  "main": "lib/FlexView.js",
```
Comment (FrancescoCioria) [inline package.json]: NOTE: if we re-add `src/index.js` (see last comment) this should become `"main": "lib",`

File: .gitignore [hunk 7]
```diff
@@ -1,6 +1,5 @@
 node_modules
 coverage
 npm-debug.log
-lib
```
Comment (FrancescoCioria) [inline .gitignore]: please revert when ready to merge

File: .npmignore [hunk 8]
```diff
@@ -1,7 +0,0 @@
-node_modules
-coverage
-npm-debug.log
-.git
-test
-karma*
-.eslintrc
```
Comment (FrancescoCioria) [inline .npmignore]: see comment on `files` addition to `package.json`

File: src/index.js (hunk 5 above)
Comment (giogonzo) [inline src/index.js]: not breaking really, the `"main"` pointer has been update in `package.json`

File: package.json (hunk 4 above)
Comment (giogonzo) [inline package.json]: is it used to import the `.scss` file?
